### PR TITLE
[inductor] Replay cond_fn mutations of captured tensors in while_loop

### LINKED ISSUE #195327: [inductor][while_loop] cond_fn-only mutation of a pre-mutated captured tensor is not replayed
I am investigating PyTorch correctness issues relevant to stateful simulation loops. I asked Codex to re-evaluate this behavior against current `main` and existing unmerged PRs. After reviewing the narrowed scope and its relationship to #184164, I decided to report it because the first compiled invocation can appear correct while leaving persistent state stale, after which later invocations silently return wrong results.

AI assistance disclosure: Codex prepared and executed the reproducer and drafted the technical analysis quoted below. I reviewed the narrowed analysis before submitting it.

> ## Bug
>
> In inference, `torch.while_loop` permits `cond_fn` and `body_fn` to mutate captured tensors that are not carried inputs. With Python sources at current `main`, Inductor fails to replay a `cond_fn`-only mutation to external storage when the captured tensor was already updated before entering the loop.
>
> https://github.com/pytorch/pytorch/blob/e4d9b6187e6ef2635cc2b648fbb409d25d6a9d9a/torch/_higher_order_ops/while_loop.py#L136-L209
>
> Eager and `aot_eager` preserve the mutation. Inductor updates the internal functional tensor, so the first call's returned clone is correct, but leaves the external tensor at the value immediately before the HOP (`0.8` on the first call). A second call then starts from that stale value and silently produces wrong numerical results.
>
> ## Reproducer
>
> ```python
> import torch
>
>
> def f(state):
>     state.mul_(0.8)
>
>     def cond(i, acc):
>         state.mul_(0.5)
>         return i < 2
>
>     def body(i, acc):
>         return i + 1, acc + state
>
>     i, acc = torch.while_loop(
>         cond,
>         body,
>         (
>             torch.zeros((), dtype=torch.int64),
>             torch.zeros_like(state),
>         ),
>     )
>     return i, acc, state.clone()
>
>
> def run(backend):
>     torch._dynamo.reset()
>     state = torch.ones((), dtype=torch.float64)
>     fn = f if backend == "eager" else torch.compile(
>         f, backend=backend, fullgraph=True
>     )
>
>     results = []
>     with torch.inference_mode():
>         for _ in range(2):
>             output = fn(state)
>             results.append(
>                 (
>                     tuple(item.item() for item in output),
>                     state.item(),
>                 )
>             )
>     print(backend, results)
>
>
> for backend in ("eager", "aot_eager", "inductor"):
>     run(backend)
> ```
>
> Each result is `(returned_values, external_state_after_call)`:
>
> ```text
> eager [
>   ((2, 0.6000000000000001, 0.1), 0.1),
>   ((2, 0.06000000000000001, 0.010000000000000002), 0.010000000000000002)
> ]
> aot_eager [
>   ((2, 0.6000000000000001, 0.1), 0.1),
>   ((2, 0.06000000000000001, 0.010000000000000002), 0.010000000000000002)
> ]
> inductor [
>   ((2, 0.6000000000000001, 0.1), 0.8),
>   ((2, 0.4800000000000001, 0.08000000000000002), 0.6400000000000001)
> ]
> ```
>
> The same failure occurs under `torch.no_grad()`. It also reproduces when the condition is false on its first evaluation: eager persists the mutation made by that one `cond_fn` call, whereas Inductor leaves external state at the pre-loop value even though `body_fn` never runs.
>
> ## Expected behavior
>
> Inductor should match eager and `aot_eager`: mutations of non-carried captured tensors made by `cond_fn` should be replayed to their external storage, including when the loop executes zero body iterations and across repeated compiled calls.
>
> ## Likely cause
>
> Dynamo already unions the storage mutations discovered in `cond_fn` and `body_fn` when constructing the higher-order operator:
>
> https://github.com/pytorch/pytorch/blob/e4d9b6187e6ef2635cc2b648fbb409d25d6a9d9a/torch/_dynamo/variables/higher_order_ops.py#L932-L987
>
> In contrast, `torch._inductor.ir.WhileLoop.create` obtains mutation indices only from `body_fn`:
>
> ```python
> mutated_idxs = check_input_alias_and_mutation(
>     body_fn.graph.module, fake_all_inputs
> )[3]
> ```
>
> https://github.com/pytorch/pytorch/blob/e4d9b6187e6ef2635cc2b648fbb409d25d6a9d9a/torch/_inductor/ir.py#L11812-L11863
>
> Consequently, a captured additional input mutated only by `cond_fn` receives no `MutationOutput`/scheduler write dependency. A local prototype that unions the `cond_fn` and `body_fn` mutation indices, on top of the additional-input handling from #184164, fixes the nonzero-iteration, zero-iteration, and repeated-call cases above.
>
> ## Relationship to #184164
>
> #184164 is a closely related partial fix, not an exact duplicate. It fixes mutation replay for captured additional inputs mutated by `body_fn`. Applying that PR locally fixes the body-mutation reproducer but does not fix this `cond_fn`-only reproducer.
>
> The PR test's cond-mutated counter is lowered from an unchanged graph-input base, while a separate buffer is mutated in `body_fn`; it therefore does not exercise a computed functionalization base mutated only by `cond_fn`. This issue can likely be handled by extending #184164 to collect mutation indices from both subgraphs and adding the two regression boundaries above.
>
> Related issue: #182036
>
> ## Versions
>
> - Python sources: current `main` at `e4d9b6187e6ef2635cc2b648fbb409d25d6a9d9a`
> - Native extension build used with that source tree: `2.15.0a0+gitb3632e5` (`b3632e56519f39da86e50fb0aa27235d626e8306`)
> - Python: 3.12.3
> - OS: Ubuntu 24.04.4 LTS, x86_64
> - CPU: Intel Core i7-6700
> - `fullgraph=True`, CPU, `torch.inference_mode()`; CUDA was not involved

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @ydwu4 @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #195393

Fixes #195327

In inference mode, torch.while_loop allows cond_fn and body_fn to
mutate captured tensors that are not carried inputs. Inductor failed to
replay a cond_fn-only mutation to external storage when the captured
tensor was mutated before entering the loop: the first compiled call
returned correct values but left the external tensor stale, and later
calls silently produced wrong results.

Root cause: after AOT, the loop is wrapped as
auto_functionalized_v2(while_loop, ..., _all_bases=[...]) where the
base of the pre-mutated captured tensor is a computed intermediate,
with a copy_ epilogue writing back to the graph input. Inductor's
post-grad decomposition back to a plain while_loop does not carry
mutation metadata on the node, so WhileLoop.create re-derives it with
check_input_alias_and_mutation -- but only scanned body_fn, and only
registered MutationOutput for carried inputs (an unreachable "only
carries can be mutated" assertion guarded the additional-input case).
With no MutationOutput, the scheduler treated the loop as a pure reader
of the base buffer and fused the copy-back into the pre-loop kernel,
writing the stale pre-loop value into the user's tensor.

The fix unions mutated input indices from both cond_fn and body_fn
(the same pattern the Switch/cond lowering uses across branches) and
registers a MutationOutput for each mutated additional input so
outer-graph reads are ordered after the loop. It also replaces a
latent iterator/index mispairing for mutated carried inputs
(check_input_alias_and_mutation returns indices in graph-node order,
not ascending order) with direct all_inputs[idx] indexing. An
alternative of plumbing mutated_arg_indices through the
auto_functionalized_v2 decomposition was rejected: the v2 schema drops
that kwarg by design, and the subgraphs' copy_ nodes are the ground
truth at lowering time. The stack_output (training) path is unaffected
because functionalization rejects subgraph input mutations there.

Test Plan:

New device-generic regression tests (CPU+CUDA) covering cond/body/both
mutation of a pre-mutated captured tensor, the zero-iteration case, and
repeated compiled calls; all 18 instantiations fail before the fix and
pass after:

```
python test/functorch/test_control_flow.py -k "pre_mutated"
```

Targeted existing suites, all passing:

```
python test/functorch/test_control_flow.py -k "auto_functionalize"
python test/functorch/test_control_flow.py -k "while_loop"
python test/inductor/test_control_flow.py -k "while"
```

Lint:

```
lintrunner -a torch/_inductor/ir.py test/functorch/test_control_flow.py
```

Authored with an AI assistant (Claude Code).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by ruddyscent** [comment]
I reviewed the diff and the following AI-assisted analysis, and I agree with its conclusion.

> AI-assisted review: I found no blocking issues in the while_loop captured-tensor mutation replay changes. The regression coverage includes cond-only, body-only, both, zero-iteration, and repeated-call cases. Focused CPU checks matched eager behavior for the PR head; the full test file was not independently run in the detached worktree.

Approving.

**Comment by anijain2305** [comment]
@claude review this PR

**Review by anijain2305 — CHANGES_REQUESTED** [review]
Can you take a look at the claude review? It points to a potential issues.

**Comment by jansel** [comment]
@claude review again

**Comment by jansel** [comment]
@claude review again, finish in under the 1 hour timeout

**Review by anijain2305 — APPROVED** [review]

**Review by ruddyscent — CHANGES_REQUESTED** [review]
I re-reviewed the updated PR and reproduced the `cpp_wrapper` failure. The new test makes carried-input mutation depend on an unrelated captured-tensor mutation, contradicting the documented contract. I therefore agree with the review below and withdraw my earlier approval.

> [AI-assisted review]
>
> ## PR Review: #195393
>
> ### Summary
>
> This PR fixes mutation replay for non-carried captured tensors in while_loop. One API-design issue blocks approval: the new test makes an otherwise rejected carried-input mutation succeed when a pre-mutated captured tensor is also mutated, and that combination fails in the C++ wrapper codegen path.
>
> ### API Design
>
> test/functorch/test_control_flow.py:13185 — Do not make carried-input mutation validity depend on an unrelated captured-tensor mutation.
>
> torch.while_loop explicitly prohibits cond_fn and body_fn from mutating carried inputs in place (https://github.com/pytorch/pytorch/blob/ee8086da674d793678bcb94cb9cb6dcee7b92fdc/torch/_higher_order_ops/while_loop.py#L192-L203). However, test_while_loop_cond_mutates_carry_and_pre_mutated_tensor (https://github.com/pytorch/pytorch/blob/ee8086da674d793678bcb94cb9cb6dcee7b92fdc/test/functorch/test_control_flow.py#L13170-L13204) requires acc.mul_() to compile successfully when the condition also mutates the pre-mutated captured tensor y. This conflicts with both the documented restriction and the adjacent test that expects carried-input mutation to be rejected.
>
> The combined case returns through the auto-functionalization path before _check_alias_and_mutation runs (https://github.com/pytorch/pytorch/blob/ee8086da674d793678bcb94cb9cb6dcee7b92fdc/torch/_higher_order_ops/while_loop.py#L621-L653). The new lowering then gives acc both a MultiOutput and a MutationOutput (https://github.com/pytorch/pytorch/blob/ee8086da674d793678bcb94cb9cb6dcee7b92fdc/torch/_inductor/ir.py#L11834-L11865). The C++ wrapper interprets any mutation output as replacing the regular carried output, leaves the MultiOutput unused, and raises while_loop has unused non-mutated outputs (https://github.com/pytorch/pytorch/blob/ee8086da674d793678bcb94cb9cb6dcee7b92fdc/torch/_inductor/codegen/cpp_wrapper_cpu.py#L2895-L2928).
>
> I reproduced this on the exact PR head with cpp_wrapper=True; both dynamic=False and dynamic=True fail with the same InductorError.
>
> Please reject carried-input mutation before the auto-functionalization split and convert the mixed case into an assertRaisesRegex regression test. If this behavior is intentionally being added as supported functionality, the documented contract and every affected wrapper path need to be updated and tested together.
>
> ### Recommendation
>
> Request Changes
>
> The PR currently establishes behavior that contradicts the public restriction and fails in another codegen path.
